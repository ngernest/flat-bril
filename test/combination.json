{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "k",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n",
            "k"
          ],
          "dest": "final",
          "funcs": [
            "combo"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "final"
          ],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "k",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "n",
            "zero"
          ],
          "dest": "nneg",
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "k",
            "zero"
          ],
          "dest": "kneg",
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "nneg",
            "kneg"
          ],
          "dest": "eitherneg",
          "op": "or",
          "type": "bool"
        },
        {
          "args": [
            "eitherneg"
          ],
          "labels": [
            "negatives",
            "natural"
          ],
          "op": "br"
        },
        {
          "label": "negatives"
        },
        {
          "args": [
            "zero"
          ],
          "op": "ret"
        },
        {
          "label": "natural"
        },
        {
          "args": [
            "k",
            "zero"
          ],
          "dest": "kzero",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "kzero"
          ],
          "labels": [
            "zeroes",
            "positive"
          ],
          "op": "br"
        },
        {
          "label": "zeroes"
        },
        {
          "args": [
            "one"
          ],
          "op": "ret"
        },
        {
          "label": "positive"
        },
        {
          "args": [
            "n",
            "k"
          ],
          "dest": "nltk",
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "nltk"
          ],
          "labels": [
            "negatives",
            "valid"
          ],
          "op": "br"
        },
        {
          "label": "valid"
        },
        {
          "args": [
            "n"
          ],
          "dest": "nfac",
          "funcs": [
            "factorial"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "k"
          ],
          "dest": "kfac",
          "funcs": [
            "factorial"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "n",
            "k"
          ],
          "dest": "nmink",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "nmink"
          ],
          "dest": "nkfac",
          "funcs": [
            "factorial"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "kfac",
            "nkfac"
          ],
          "dest": "comboden",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "nfac",
            "comboden"
          ],
          "dest": "combo",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "combo"
          ],
          "op": "ret"
        }
      ],
      "name": "combo",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "x",
            "one"
          ],
          "dest": "basee",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "basee"
          ],
          "labels": [
            "base",
            "recurs"
          ],
          "op": "br"
        },
        {
          "label": "base"
        },
        {
          "args": [
            "one"
          ],
          "op": "ret"
        },
        {
          "label": "recurs"
        },
        {
          "args": [
            "x",
            "one"
          ],
          "dest": "oneless",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "oneless"
          ],
          "dest": "stepback",
          "funcs": [
            "factorial"
          ],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x",
            "stepback"
          ],
          "dest": "fact",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "fact"
          ],
          "op": "ret"
        }
      ],
      "name": "factorial",
      "type": "int"
    }
  ]
}
