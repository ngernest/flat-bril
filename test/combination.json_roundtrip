{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "k",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n",
            "k"
          ],
          "dest": "final",
          "funcs": [
            "combo"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "final"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "k",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "n",
            "zero"
          ],
          "dest": "nneg",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "k",
            "zero"
          ],
          "dest": "kneg",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "nneg",
            "kneg"
          ],
          "dest": "eitherneg",
          "funcs": [],
          "labels": [],
          "op": "or",
          "type": "bool"
        },
        {
          "args": [
            "eitherneg"
          ],
          "funcs": [],
          "labels": [
            "negatives",
            "natural"
          ],
          "op": "br"
        },
        {
          "label": "negatives"
        },
        {
          "args": [
            "zero"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "natural"
        },
        {
          "args": [
            "k",
            "zero"
          ],
          "dest": "kzero",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "kzero"
          ],
          "funcs": [],
          "labels": [
            "zeroes",
            "positive"
          ],
          "op": "br"
        },
        {
          "label": "zeroes"
        },
        {
          "args": [
            "one"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "positive"
        },
        {
          "args": [
            "n",
            "k"
          ],
          "dest": "nltk",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "nltk"
          ],
          "funcs": [],
          "labels": [
            "negatives",
            "valid"
          ],
          "op": "br"
        },
        {
          "label": "valid"
        },
        {
          "args": [
            "n"
          ],
          "dest": "nfac",
          "funcs": [
            "factorial"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "k"
          ],
          "dest": "kfac",
          "funcs": [
            "factorial"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "n",
            "k"
          ],
          "dest": "nmink",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "nmink"
          ],
          "dest": "nkfac",
          "funcs": [
            "factorial"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "kfac",
            "nkfac"
          ],
          "dest": "comboden",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "nfac",
            "comboden"
          ],
          "dest": "combo",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "combo"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "combo",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "x",
            "one"
          ],
          "dest": "basee",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "basee"
          ],
          "funcs": [],
          "labels": [
            "base",
            "recurs"
          ],
          "op": "br"
        },
        {
          "label": "base"
        },
        {
          "args": [
            "one"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "recurs"
        },
        {
          "args": [
            "x",
            "one"
          ],
          "dest": "oneless",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "oneless"
          ],
          "dest": "stepback",
          "funcs": [
            "factorial"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "x",
            "stepback"
          ],
          "dest": "fact",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "fact"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "factorial",
      "type": "int"
    }
  ]
}
