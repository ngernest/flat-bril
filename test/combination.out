@main
	op: call 		dest: final		type: int		args: ["n", "k"]		funcs: ["combo"]
	op: print		args: ["final"]	
@combo
	op: const		dest: zero		type: int		value: 0	
	op: const		dest: one		type: int		value: 1	
	op: lt   		dest: nneg		type: bool		args: ["n", "zero"]	
	op: lt   		dest: kneg		type: bool		args: ["k", "zero"]	
	op: or   		dest: eitherneg		type: bool		args: ["nneg", "kneg"]	
	op: br   		args: ["eitherneg"]		labels: ["negatives", "natural"]
.negatives
	op: ret  		args: ["zero"]	
.natural
	op: eq   		dest: kzero		type: bool		args: ["k", "zero"]	
	op: br   		args: ["kzero"]		labels: ["zeroes", "positive"]
.zeroes
	op: ret  		args: ["one"]	
.positive
	op: lt   		dest: nltk		type: bool		args: ["n", "k"]	
	op: br   		args: ["nltk"]		labels: ["negatives", "valid"]
.valid
	op: call 		dest: nfac		type: int		args: ["n"]		funcs: ["factorial"]
	op: call 		dest: kfac		type: int		args: ["k"]		funcs: ["factorial"]
	op: sub  		dest: nmink		type: int		args: ["n", "k"]	
	op: call 		dest: nkfac		type: int		args: ["nmink"]		funcs: ["factorial"]
	op: mul  		dest: comboden		type: int		args: ["kfac", "nkfac"]	
	op: div  		dest: combo		type: int		args: ["nfac", "comboden"]	
	op: ret  		args: ["combo"]	
@factorial
	op: const		dest: one		type: int		value: 1	
	op: eq   		dest: basee		type: bool		args: ["x", "one"]	
	op: br   		args: ["basee"]		labels: ["base", "recurs"]
.base
	op: ret  		args: ["one"]	
.recurs
	op: sub  		dest: oneless		type: int		args: ["x", "one"]	
	op: call 		dest: stepback		type: int		args: ["oneless"]		funcs: ["factorial"]
	op: mul  		dest: fact		type: int		args: ["x", "stepback"]	
	op: ret  		args: ["fact"]	
